DESIGN.md

```diff
@@ -0,0 +1,127 @@
+# Design rationale for SCIP
+
+Sourcegraph supports viewing and navigation code.
+This is similar to an IDE, except for the actual "editor" part.
+
+The design of SCIP is based on this primary motivating use case,
+as well as for fixing the pain points we encountered with using LSIF.[^1]
+
+SCIP is meant to be _transmission_ format for sending
+data from some producers to some consumers
+-- it is not meant as a _storage_ format for querying.
+Ideally, producers should be able to directly output SCIP
+instead of going through an intermediate format
+and then converting to SCIP for transmission.
+
+[^1]: Sourcegraph historically supported LSIF uploads
+as well as maintained our own LSIF indexers,
+but we ran into issues of development velocity,
+debugging, as well as indexer performance bottlenecks.
+
+## Goals
+
+### Primary goals
+
+- Support code navigation at the fidelity of state-of-the-art IDEs.
+  - Why: We want people to have an excellent experience navigating
+    their code inside Sourcegraph itself.
+- Ease of writing indexers (i.e. producers).
```

Comment:
Worth mentioning across multiple language ecosystems (hence protobuf instead of more obscure binary format)